R/rnet-clean.R

```diff
@@ -80,61 +84,174 @@
-    rnet_breakup_collection <- lwgeom::st_split(rnet, intersection_points$geometry)
-    rnet_clean <- sf::st_collection_extract(rnet_breakup_collection, "LINESTRING")
-  } else {
-    rnet_clean <- rnet
+  # Check if we need to rebuild tbl_df at the end
+  rebuild_tbl <- FALSE
+  if (inherits(rnet, "sf") && inherits(rnet, "tbl_df")) {
+    rebuild_tbl <- TRUE
   }
 
-  # Split again at the duplicated internal vertexes
-  rnet_internal_vertexes_duplicated <- rnet_internal_vertexes[duplicated(rnet_internal_vertexes)]
+  # step 1 - Find points that are both boundary points and internal points.
+  # 1a - Find internal points (i.e. points that are not in the boundary)
+  internal_points <- data.table::data.table(
+    sfheaders::sf_to_df(line2vertices(rnet))
+  )
+  data.table::setkey(internal_points, x, y)
+  # I don't need duplicated pairs of coordinates
+  internal_points <- unique(internal_points)
+  if (verbose) {
+    message("Extracted the (unique) internal points")
+  }
+  # 1b - Find points that lie in the boundaries
+  boundary_points <- data.table::data.table(
+    sfheaders::sf_to_df(rnet_boundary_unique(rnet))
+  )
+  data.table::setkey(boundary_points, x, y)
+  if (verbose) {
+    message("Extracted the (unique) boundary points")
+  }
 
-  if (length(rnet_internal_vertexes_duplicated) > 0 & breakup_internal_vertex_matches) {
-    message("Splitting rnet object at the shared internal points.")
-    rnet_breakup_collection <- lwgeom::st_split(rnet_clean, rnet_internal_vertexes_duplicated)
-    rnet_clean <- sf::st_collection_extract(rnet_breakup_collection, "LINESTRING")
+  # Consider only the points that are both internal and boundaries
+  shared_internal_points_data_table <- boundary_points[
+    # The following [ ] is used to filter the boundary points which are also
+    # internal points. The operation checks that the x and y coordinates
+    # are identical
+    stats::na.omit(boundary_points[internal_points, which = TRUE]),
+  ]
+
+  if (nrow(shared_internal_points_data_table) > 0) {
+    # step 2 - Split at shared internal points
+    message("Splitting the input object at the shared internal points.")
+    rnet <- my_st_split(rnet, shared_internal_points_data_table)
   }
 
-  rnet_clean
+  # step 3 - Find duplicated internal points
+  internal_points <- data.table::data.table(
+    sfheaders::sf_to_df(line2vertices(rnet))
+  )
+  data.table::setkey(internal_points, x, y)
+  duplicated_internal_points <- internal_points[duplicated(internal_points)]
+
+  # 4 - Split at duplicated internal points
+  if (nrow(duplicated_internal_points) > 0) {
+    message("Splitting rnet object at the duplicated internal points.")
+    rnet <- my_st_split(rnet, duplicated_internal_points)
+  }
+
+  # 5 - Maybe we need to rebuild the tbl_df structure
+  if (rebuild_tbl) {
+    rnet <- sf::st_as_sf(dplyr::as_tibble(rnet))
+  }
+
+  rnet
+}
+
+# Not sure if it should be exported (maybe yes). In any case: ADD DOCS
+# FIXME: ADD DOCS
```

Comment:
No harm in exporting it, it's a useful function.

Reply:
👍

Reply:
Still need to add tests/docs